Feature/add file stream input handling

maciej-kulon opened the pull request:
This PR will add local files handling and fix issue with HTTP errors not being thrown.

maciej-kulon committed:
feat(): add local file stream

maciej-kulon committed:
fix(): http errors were not thrown

Batuzz commented on src/errors.ts:
```diff
@@ -0,0 +1,17 @@
+export class MediaInfoInputError extends MediaInfoError {
+  constructor(
+    public readonly message: string,
+    public readonly input: PathLike,
```
I guess it should be `MediaInfoInput` type?

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
+  private static openFileStream(input: string) {
+    try {
+      const stream = fs.createReadStream(input);
+      if(stream.errored) {
```
```suggestion
      if (stream.errored) {
```

Batuzz replied:
please run linter to check if there any other places that do not satisfy linter

maciej-kulon replied:
Done 😄

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
+      }
+      return stream;
+    } catch (err) {
+      throw new MediaInfoInputError('Could not open read stream to the file.', input, err);
```
```suggestion
      throw new MediaInfoInputError('Could not open read stream to the file', input, err);
```

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
+      }
+      return stream;
+    } catch (err) {
+      throw new MediaInfoInputError('Could not open read stream from HTTP address.', input, err);
```
```suggestion
      throw new MediaInfoInputError('Could not open read stream from HTTP address', input, err);
```

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
+  private static openHttpStream(input: string) {
+    try {
+      const stream = got.stream(input);
+      if(stream.errored) {
```
```suggestion
      if (stream.errored) {
```

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
+      stream = this.openFileStream(input);
     } else {
-      throw new Error('Input other than HTTP is not supported yet');
+      throw new MediaInfoInputError('Input is neither HTTP address nor file path.', input);
```
```suggestion
      throw new MediaInfoInputError('Input is neither HTTP address nor file path', input);
```

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
     if (input.toLowerCase().startsWith('http')) {
-      stream = got.stream(input);
+      stream = this.openHttpStream(input);
+    } else if (fs.existsSync(input)) {
```
I wonder if we should use sync methods - I guess we shouldn't

Batuzz replied:
from [StackOverflow](https://stackoverflow.com/questions/17699599/node-js-check-if-file-exists):

Edit: Since node v10.0.0we could use [fs.promises.access(...)](https://nodejs.org/api/fs.html#fs_fspromises_access_path_mode)

Example async code that checks if file exists:

```
function checkFileExists(file) {
  return fs.promises.access(file, fs.constants.F_OK)
           .then(() => true)
           .catch(() => false)
}
```

maciej-kulon replied:
We do not need that. I've removed check if file exists. If file does not exists it will simply throw an error with such information.

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
     }
- 
+    
+    if(!stream) {
```
```suggestion
    if (!stream) {
```

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +47,45 @@ export class MediaInfo {
- 
+    
+    if(!stream) {
+      errorHandler(`Could not create any stream for a given input: ${input}`);
```
maybe `new Error('msg')`? also, I wonder what would need to happen in order to reach this code

maciej-kulon replied:
Gonna remove that - it was a part of the investigation with HTTP errors problem, it does nothing at this point 😄

Batuzz commented on src/index.ts:
```diff
@@ -72,11 +97,9 @@ export class MediaInfo {
       stream.on('data', (chunk) => {
         mediaInfoInstance.Open_Buffer_Continue(chunk);
-        seekTo = mediaInfoInstance.Open_Buffer_Continue_Goto_Get();
+        mediaInfoInstance.Open_Buffer_Continue_Goto_Get();
```
why the `seekTo` was not required?

maciej-kulon replied:
Because the only purpose of that was to check if its value is -1 or not, depending on that we were calling stream.end() function. Thats not the case anymore - as eslint noticed 😄

maciej-kulon committed:
refactor(): add input handlers

- move input type handling responsibility to external modules
- add common interface for input handling objects

Batuzz commented on src/index.ts:
```diff
@@ -51,16 +46,10 @@ export class MediaInfo {
-      throw new Error('Input other than HTTP is not supported yet');
-    }
- 
+  private static async getDataStream(input: string, errorHandler: ErrorHandlerFunction) {
```
would be nice to have return type

maciej-kulon committed:
refactor(): remove async func from promise

Batuzz commented on src/inputHandlers/fileInputHandler.ts:
```diff
@@ -0,0 +1,17 @@
+import * as fs from 'fs';
```
```suggestion
import { ReadStream, createReadStream } from 'fs';
```

Batuzz commented on src/inputHandlers/httpInputHandler.ts:
```diff
@@ -0,0 +1,18 @@
+import got from 'got/dist/source';
```
`* as got from got`?

maciej-kulon replied:
Done.

maciej-kulon committed:
chore(): comments resolving

- add return type to getDataStream function
- change import in fileInputHandler

maciej-kulon committed:
chore(): import fix
